pkg/gatewayserver/gatewayserver.go

```diff
@@ -654,19 +654,24 @@ func (gs *GatewayServer) handleUpstream(conn connectionEntry) {
 	}
 }
 
-// UpdateConnectionStats updates the stats for a single gateway connection.
-func (gs *GatewayServer) UpdateConnectionStats(ctx context.Context, conn *io.Connection) error {
-	return gs.statsRegistry.Set(ctx, conn.Gateway().GatewayIdentifiers, conn.Stats())
-}
-
 func (gs *GatewayServer) updateConnStats(conn connectionEntry) {
 	ctx := conn.Context()
 	logger := log.FromContext(ctx)
 
+	// Initial dummy update, so that gateway appears connected
+	if err := gs.UpdateConnectionStats(conn.Connection, false, false, true); err != nil {
+		logger.WithError(err).Error("Failed to initialize connection stats")
+	}
+
+	var (
+		lastUpdateUp, lastUpdateDown, lastUpdateStatus time.Time
+		newUp, newDown, newStatus                      bool
+	)
+
 	defer func() {
 		logger.Debug("Delete connection stats")
-		err := gs.statsRegistry.Set(ctx, conn.Gateway().GatewayIdentifiers, nil)
-		if err != nil {
+
+		if err := gs.ClearConnectionStats(conn.Connection, newUp, newDown, true); err != nil {
```

Comment:
What I still don't entirely understand, is that status is deleted regardless of whether this connection ever handled a status message? Shouldn't that be only if there has been a status message at all here?

Reply:
The `status` is used to keep the `ConnectedAt` and `Protocol` fields as well. If we do not delete this on disconnect, then the gateway will still show up as connected.

Reply:
OK well I guess that is the way to go then